### ISSUE TITLE
tracing-subscriber: (minimal) no-std support

### ISSUE THREAD

**Issue by hawkw** [tokio-rs/tracing#999]
### Crates

- `tracing-subscriber`

### Motivation
Presently, the `tracing-subscriber` crate requires the Rust standard library and doesn't build with `#![no-std]` targets. For the most part, this is fine, as much of `tracing-subscriber` inherently depends on `std` APIs. 

However, `tracing-subscriber` also contains some key abstractions that are necessary for interoperability: the `Layer` and `LookupSpan` traits. Since these traits are in `tracing-subscriber`, `no-std` users cannot currently access them.

### Proposal

We should add a `std` feature flag that defaults to on, and expose a minimal API when it is disabled. At the very least, we can expose the `Layer` and `LookupSpan` traits. 

This would mean that, even though the `tracing-subscriber` `Registry` type will not support `no-std`, `no-std` users can build their own custom registry and use the same `Layer` abstraction as the rest of the world. The correct registry behavior is likely to be use-case and platform-dependent, so it's fine for no-std users to have to build their own alternative registries, but exposing `Layer` should at least allow them to re-use more general purpose subscriber components.

Since the next `tracing-subscriber` release will be breaking, it's fine to add a new default-on feature flag for `std`.

### Alternatives

- We could move the traits to a separate crate. I'd rather not do this, since it increases the (already quite large) number of crates we have to version and release.
- Not support `no-std` in `tracing-subscriber`. This would be a bummer: `tracing` and `tracing-core` _do_ have `no-std` support, and not supporting it in the core `tracing-subscriber` abstractions would lock `no-std` users out of a majority of the tracing ecosystem.

**Comment by hawkw** [comment]
cc @hoverbear

**Comment by Hoverbear** [comment]
I've definitely bumped into this in Embedded work!

Would the scope of this be to just flag out anything not viable on `core+alloc` until a future version?

**Comment by hawkw** [comment]
> Would the scope of this be to flag out anything not viable on `core+alloc` until a future version?

Yup, that's right. We should definitely be able to expose the `Layer` and `LookupSpan` traits, at the very least.
